### ISSUE TITLE
"NewType" has bad repr

### ISSUE THREAD

**Issue by hoodmane** [davidhalter/jedi#1652]
This is really a problem with the `typing` package, [see the typing issue I opened](https://github.com/python/typing/issues/746), but it could be fixed in Jedi too.
```python
import inspect
from typing import NewType
A = NewType("A", int)
def test(a : A):
    pass
sig = inspect.signature(test)
annotation = list(sig.parameters.values())[0].annotation
print(inspect.formatannotation(annotation))
```
The result is:
```python
<function NewType.<locals>.new_type at MEMORYADDRESS>
```
This directly shows up in Jedi, so the following code:
```python
import jedi
from typing import NewType

A = NewType("A", int)

def test(a : A):
    pass

test_jedi = jedi.Interpreter("test", [globals()]).complete()[0]
print(test_jedi.docstring())
```
outputs
```python
test(a: <function NewType.<locals>.new_type at MEMORYADDRESS>)
```
I would rather if the output were:
```python
test(a: A)
```
The source for `inspect.formatannotation` is quite managable:

```python
def formatannotation(annotation, base_module=None):
    if getattr(annotation, '__module__', None) == 'typing':
        return repr(annotation).replace('typing.', '') # <== the path taken in NewType case.
    if isinstance(annotation, type):
        if annotation.__module__ in ('builtins', base_module):
            return annotation.__qualname__
        return annotation.__module__+'.'+annotation.__qualname__
    return repr(annotation)
```
Maybe Jedi could use the following patched version, at least until the `typing` module fixes this issue:
```python
def formatannotation(annotation, base_module=None):
    if hasattr(annotation, "__qualname__") and annotation.__qualname__ == 'NewType.<locals>.new_type':
        return annotation.__name__
    if getattr(annotation, '__module__', None) == 'typing':
        return repr(annotation).replace('typing.', '') # <== the path taken in NewType case.
    if isinstance(annotation, type):
        if annotation.__module__ in ('builtins', base_module):
            return annotation.__qualname__
        return annotation.__module__+'.'+annotation.__qualname__
    return repr(annotation)
```

Here's a complete example monkey patching my modified `formatannotation` over `inspect.formatannotation`:
```python
import inspect
# Patch formatannotation, must be done before importing jedi!
def formatannotation(annotation, base_module=None):
    if hasattr(annotation, "__qualname__") and annotation.__qualname__ == 'NewType.<locals>.new_type':
        return annotation.__name__
    if getattr(annotation, '__module__', None) == 'typing':
        return repr(annotation).replace('typing.', '')
    if isinstance(annotation, type):
        if annotation.__module__ in ('builtins', base_module):
            return annotation.__qualname__
        return annotation.__module__+'.'+annotation.__qualname__
    return repr(annotation)
inspect.formatannotation = formatannotation

from typing import NewType
A = NewType("A", int)
def test(a : A):
    pass

import jedi
test_jedi = jedi.Interpreter("test", [globals()]).complete()[0]
print(test_jedi.docstring())
```
This has the desired output:
```python
test(a: A)
```

**Comment by davidhalter** [comment]
Have you tried to push this in CPython? IMO that would be the right place for such a change.

**Comment by hoodmane** [comment]
I'm discussing with the CPython folks in the issue I opened there. It seems that I might be persuading them. 

Any change they make won't appear in CPython until at least version 3.10. (The prerelease for 3.9 is already stabilized so I guess it's too late to make it into that version). So it could be useful for Jedi to patch in the fix for earlier versions.

**Comment by davidhalter** [comment]
It might be backported.

I guess I'll wait until they sort it out. Can you post the link to that discussion?

**Comment by hoodmane** [comment]
The discussion on CPython is here:
https://github.com/python/typing/issues/746

I would not bet on them backporting any fix, but it couldn't hurt to ask.

**Comment by davidhalter** [comment]
I'm just waiting for a decision. If they don't backport, we can do that as well.

**Comment by davidhalter** [comment]
BTW: Just if you want to not be annoyed by this for now, use `from __future__ import annotations`.

**Comment by hoodmane** [comment]
> BTW: Just if you want to not be annoyed by this for now, you can use from __future__ import annotations.

Thanks for the suggestion, this seems like a pretty good solution. One point in favor of them backporting the change to `NewType` is that in version 3.10 the `annotations` feature will be on by default and so fixing the `NewType` repr will not matter so much.

**Comment by davidhalter** [comment]
I agree. I'm closing this in favor of the 3.10 / annotations solution.
